connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala

```diff
@@ -94,83 +94,115 @@ object KustoWriter {
+      }
 
       kustoClient.setMappingOnStagingTableIfNeeded(stagingTableIngestionProperties, tableCoordinates.database, tmpTableName, table, crp)
+      if (stagingTableIngestionProperties.flushImmediately) {
+        KDSU.logWarn(myName, "It's not recommended to set flushImmediately to true")
+      }
 
-//    TODO remove until batching policy problem is good
-//      if (stagingTableIngestionProperties.flushImmediately) {
-//        KDSU.logWarn(myName, "It's not recommended to set flushImmediately to true on production")
-//      }
+      //    TODO remove until batching policy problem is good
+      //      if (stagingTableIngestionProperties.flushImmediately) {
+      //        KDSU.logWarn(myName, "It's not recommended to set flushImmediately to true on production")
+      //      }
 
       val rdd = data.queryExecution.toRdd
       val partitionsResults = rdd.sparkContext.collectionAccumulator[PartitionResult]
       if (writeOptions.isAsync) {
         val asyncWork = rdd.foreachPartitionAsync { rows => ingestRowsIntoTempTbl(rows, batchIdIfExists, partitionsResults) }
         KDSU.logInfo(myName, s"asynchronous write to Kusto table '$table' in progress")
         // This part runs back on the driver
-        asyncWork.onSuccess {
-          case _ => kustoClient.finalizeIngestionWhenWorkersSucceeded(
-            tableCoordinates, batchIdIfExists, kustoClient.engineClient, tmpTableName, partitionsResults,
-            writeOptions, crp, tableExists, rdd.sparkContext, authentication)
-        }
-        asyncWork.onFailure {
-          case exception: Exception =>
-            kustoClient.cleanupIngestionByProducts(
-              tableCoordinates.database, kustoClient.engineClient, tmpTableName, crp)
-            KDSU.reportExceptionAndThrow(myName, exception, "writing data", tableCoordinates.clusterUrl, tableCoordinates.database, table, shouldNotThrow = true)
-            KDSU.logError(myName, "The exception is not visible in the driver since we're in async mode")
+        if (writeOptions.isTransactionalMode) {
+          asyncWork.onSuccess {
+            case _ => kustoClient.finalizeIngestionWhenWorkersSucceeded(
+              tableCoordinates, batchIdIfExists, kustoClient.engineClient, tmpTableName, partitionsResults,
+              writeOptions, crp, tableExists, rdd.sparkContext, authentication)
+          }
+          asyncWork.onFailure {
+            case exception: Exception =>
+              if (writeOptions.userTempTableName.isEmpty) {
+                kustoClient.cleanupIngestionByProducts(
+                  tableCoordinates.database, kustoClient.engineClient, tmpTableName, crp)
+              }
+              KDSU.reportExceptionAndThrow(myName, exception, "writing data", tableCoordinates.clusterUrl, tableCoordinates.database, table, shouldNotThrow = true)
+              KDSU.logError(myName, "The exception is not visible in the driver since we're in async mode")
```

Comment:
I know it was already like this, but do we want a separate error trace for this? I'd think it'd be better to consolidate both lines.

Reply:
I do want it in a separate trace - this should raise awareness for why couldn't the user handle the error.. this is printed at most once in a run - why does it matter?

Reply:
NBD
But if the 2nd trace is only explaining the 1st, I'd think it's confusing to separate them.

But you're saying the 2nd trace is relevant to any exception thrown during that run?

Reply:
Yes